ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py

```diff
@@ -0,0 +1,516 @@
+        if params is None:
+            params = {}
+
+        url = self._apply_params(url, params)
+
+        semaphore = getDeferredSemaphore(self.auth_url, MAX_PARALLEL)
+
+        log.debug("GET %s", url)
+        headerargs = " ".join(["-H \"%s: %s\"" % (x[0], x[1][0]) for x in headers.getAllRawHeaders()])
+        log.debug('curl "{url}" {headerargs}'.format(
+            url=url,
+            headerargs=headerargs
+        ))
+
+        try:
+            response = yield semaphore.run(
+                add_timeout,
+                agent.request(
+                    'GET',
+                    url,
+                    headers=headers
+                ),
+                READ_TIMEOUT
+            )
+        except TimeoutError:
+            raise TimeoutError("GET %s" % url)
+        body = yield readBody(response)
+
+        # If the request resulted in an error, raise an exception
+        self.handle_error_response(response, body)
+
+        returnValue((body, response.headers))
+
+    @inlineCallbacks
+    def authenticated_GET_request(self, url, headers=None, params=None):
+        if headers is None:
+            headers = self.default_headers()
+
+        if self.token_id is None:
+            yield self.authenticate()
+        headers.setRawHeaders('X-Auth-Token', [self.token_id])
+
+        try:
+            result = yield self.GET_request(url, headers=headers, params=params)
+            returnValue(result)
+        except UnauthorizedError:
+            log.info("Received unauthorized response to http request, re-authenticating.  (%s)", url)
+            yield self.authenticate()
+            headers.setRawHeaders('X-Auth-Token', [self.token_id])
+
+        result = yield self.GET_request(url, headers=headers, params=params)
+        returnValue(result)
+
+    @inlineCallbacks
+    def POST_request(self, url, headers=None, body=None):
+        scheme = urlparse(url).scheme
+        agent = self.agent(scheme)
+        if headers is None:
+            headers = self.default_headers()
+
+        if body is None:
+            body = ""
+
+        semaphore = getDeferredSemaphore(self.auth_url, MAX_PARALLEL)
+
+        log.debug("POST %s", url)
+
+        try:
+            response = yield semaphore.run(
+                add_timeout,
+                agent.request(
+                    'POST',
+                    url,
+                    headers=headers,
+                    bodyProducer=StringProducer(body)
+                ),
+                READ_TIMEOUT
+            )
+        except TimeoutError:
+            raise TimeoutError("POST %s" % url)
+
+        body = yield readBody(response)
+
+        # If the request resulted in an error, raise an exception
+        self.handle_error_response(response, body)
+
+        returnValue((body, response.headers))
+
+    @inlineCallbacks
+    def authenticated_POST_request(self, url, headers=None, body=None):
+        if self.token_id is None:
+            yield self.authenticate()
```

Comment:
I notice in `authenticated_GET_request`, we have to put the token in the headers.  Why don't we need that for POST?  Other than that, the logic is nearly identical.  Would it make sense to have an authentication wrapper we call for both?

If this is supposed to be different, maybe we should keep them separate.  On the other hand, if it's just something that got added to one, but not the other, it's an example of the dangers of duplicating code.